api/envoy/api/v2/listener/udp_listener_config.proto

```diff
@@ -30,6 +30,17 @@ message UdpListenerConfig {
 
     google.protobuf.Any typed_config = 3;
   }
+  
+  enum UdpShutDownMode {
+    // Stop reading and send public resets to active sessions.
+    GRACEFUL_SHUTDOWN = 1;
+    // Stop processing udp immediately.
+    HARD_SHUTDOWN = 2;
+  }
+
+  UdpShutDownMode shutdown_mode = 4 [(validate.rules).enum = {defined_only: true}];
+
+  bool udp_enabled = 5;
```

Comment:
Sorry I'm a little confused about what is being proposed here. Do you mind adding some more comments?

Reply:
+!

Reply:
@mattklein123 @htuch for turning off/on udp/quick processing, my proposal is to have boolean switch on udp listener config + shutdown mode field (for shutdown type). Submitted this draft config to get community agreement on the approach, code will follow.

Reply:
Why don't we just remove the listener and drain as normal? Trying to grok what is distinct here vs. any other listener that comes to life at some point, starts accepting, and then is removed and drained.

Reply:
@htuch thanks for suggestion, i did not work much with LDS, so missed that opportunity. Agree that this is cleaner way, however currently listener manager supports only draining (graceful shutdown) of active listeners. In case we go with LDS approach, it will require code change to support multiple draining modes, right?

Reply:
Yes, but I think a drain timeout could subsume this; if you provide a drain timeout and make it zero, that will give you an immediate drain, > 0 will give you a graceful drain.

Reply:
@htuch i see. But if to read [issue](https://github.com/envoyproxy/envoy/issues/9230) description, it says `Current QUIC support can only be turned on/off via LDS. We can do better here by adding an Envoy reloadable feature for QUIC to turn the switch via runtime dynamic config push.` As i understood it meant introducing config switch.

Reply:
@htuch if we have UDP proxy and QUIC share a listener (443, select which via VIP) and want to turn off one and not the other, how do you do that without explicit config?  If there's existing framework to do that than I'm good.

Reply:
I guess this comes down to documentation. What is this `udp_enabled` supposed to do?  Does it do both UDP and QUIC? What does the drain do? Is it just for UDP or also for QUIC? Is there a parallel way to achieve the same ends for QUIC? 

Maybe @danzh2010 can weight in on what the goal is.

Reply:
Would it be better to just have a per-UDP listener that has a runtime kill switch that will just stop accepting packets? I think maybe this is a simpler way of handling the intention?

Reply:
I guess it depends how we structure shared listeners.
I thought if we had QUIC and QUIC-proxy listening on UDP/443 they'd share a listener, and some higher level filter would determine which packets went where.  In that case as said it'd be nice to able to disable them one at a time.  We can deal without but it's a nice to have.
If we think we'd have a QUIC/443 listener and a QUIC-proxy/443 listener just sharing incoming traffic via some combination of SO_REUSEPORT and BPF this would be fine.
I'd assumed we were going to do more of the former, at which point a non-listener kill switch has value IMO (though again we can live without it)

Reply:
My actual assumption (potentially naive) was that QUIC and QUIC-proxy would probably be different deployments of the binary in different auto scaling groups, etc. What is the use case in which you envision the same process doing both?

Reply:
Generic example is for edge proxies with low-value certs that do transparent proxying for latency gains.  QUIC-terminate locally served content, UDP proxy where you'd TCP proxy because DNS SRV doesn't work.
We have two different frontline deployments which do TLS / TCP proxy based on VIP and one of them did QUIC+Udp-proxy for years.
We haven't had an issue where we needed to turn off one and not the other yet, but as tunneling over QUIC picks up steam (and that team has agreed to OSS once we land core QUIC) I think it gets more useful to turn different modes down separately.
We can always start simple and add more later though!

Reply:
> We can always start simple and add more later though!

That would be my feeling. Right now we don't have any filter chain concept for UDP (I would like to fix that as part of a general listener API cleanup), and without that I'm not sure how we would do something like ^ today. IMO right now I think we could just have a per-listener runtime kill switch for UDP listeners? Or even just for the QUIC listener config itself?

Reply:
Thanks for feedback, it's valuable information to get familiar with. So we agreed to introduce separate config switches for upd and quic listeners.